### DISCUSSION TITLE
SV06 - Klipper creates bed mesh before printing, but ignores it

### DISCUSSION THREAD

**Discussion by JoernSchoenyan** [bassamanator/Sovol-SV06-firmware#114] [General]
Hello everyone,

I'm not sure where to begin. Bassam suggested on Reddit (I'm Zilla85 over there) to open a discussion on my problem.

I have a SV06 and use an old mini computer with an Atom Z8350 and 4 GB RAM to run Klipper. I've installed Klipper, Moonraker and Mainsail (running Debian) an the computer, using the instructions from Mainsail's manual setup guide. Then, I've pulled Bassam's config, enabled the KAMP variable and configured my slicer (I use Cura, it is the only slicer I'm comfortable with).

This setup kind of works, it creates a bed mesh before every print and makes a nice little purge line, so I can be 100% sure KAMP is enabled. But the bed mesh is completly ignored. I have a bed level variance of ~0.175mm, my first layer is consistently bad and I can exactly see where the bed is a bit higher or lower, this fits perfectly to the bed mesh.

The SV06 is nearly stock. I've assembled it a few months ago and the stock firmware worked okay-ish. I have a thermal insulation under the heat bed and after installing Klipper, I've realized that my bed level variance was like 0.6mm, so I did that silicone tube mod, which improved the variance to the already mentioned ~0.175mm. 

Any ideas what causes Klipper to create a bed mesh and then ignoring it? What informations do you need?

**Comment by JoernSchoenyan** [comment]
I did some more tests, I've disabled KAMP and edited my slicer config to load the bed mesh instead before printing, still ABL is ignored.

**Comment by bassamanator** [comment]
Hi @JoernSchoenyan,

Let me see see that if I have this straight.
- The klipper install worked.
- The printer is printing.
- Before every print, there is a KAMP mesh (localized to the print area) being created because you see it happen.

The **problem** is that even though there is a mesh created, your first layers are quite bad?

**Comment by bassamanator** [comment]
> I did some more tests, I've disabled KAMP and edited my slicer config to load the bed mesh instead before printing, still ABL is ignored.

So firstly, it's not possible to ignore the mesh, that is how I wrote the `print_start` macro.

> But the bed mesh is completly ignored. I have a bed level variance of ~0.175mm, my first layer is consistently bad and I can exactly see where the bed is a bit higher or lower, this fits perfectly to the bed mesh.

The printer should be able to handle a variance of 0.175mm. Please print [this part](https://www.printables.com/model/468179-bed-level-bed-mesh-calibration-sv06) and tell me where the layer is bad. I already have my suspicions but I'll wait.

Please also post a picture oriented as follows:
![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/61985779/a2a7258a-22ce-469e-a216-d0324ba3914f)

**Reply by JoernSchoenyan** [reply]
Printed as like the bed mesh isn't used, but I agree this shouldn't be possible with your macro at all. I begin to suspect this is a Klipper bug.

![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/1822665/7398a1ad-5203-4355-84fe-1c292f29857e)
![IMG_20231126_125757](https://github.com/bassamanator/Sovol-SV06-firmware/assets/1822665/0ad90823-e8f8-431f-a567-3955c37bf437)

**Comment by bassamanator** [comment]
This is almost what I was expecting, nozzle too close to the bed on the left, nozzle too far from the bed, or just right, on the right side of the bed.

This is not a bug, this is a physical problem with the alignment of the printer.

1. Disable KAMP. It's going to complicate everything. Once the problem is fixed, you can re-enable KAMP.
2. Run through all the initial steps found at https://github.com/bassamanator/everything-sovol-sv06, especially the gantry squaring video.

Take your time with getting things on that list correct. Once ready, create another mesh, and do another bed-leveling print and report back.

**Comment by JoernSchoenyan** [comment]
Okay, here I am to report back. It's very difficult. First things first: yes, the gantry wasn't perfectly square. I did assemble it with a set square, but I think it was a tiny bit loose and moved when it was warm in the summer or something.

I disassembled the printer, cleaned it, took the opportunity to grease the bearings the proper way, put it together again as perfect and square as I could. Then, I've put more effort into the silicone tube mod and modified it for my heatbed, now I'm using only silicone tubes at the corner, the middle is now using the stock spacer and this way the bed is a bit more flat (0.16mm now). Long story short: even then, after calibration, the print was bad, it looked the same like the test print above and I aborted it.

To double check, I've flashed the original firmware on the printer, calibrated it and the print isn't perfect, but pretty good. And I can see the Z axis moving while printing to compensate the bed variance which I can't see on Klipper. Now I think about if it is worth going down the rabbit hole and reporting a bug over at Klipper, but I don't really know.

**Reply by Notaseraf** [reply]
Try x-twist calibration? 
https://github.com/bassamanator/Sovol-SV06-firmware/discussions/81

**Comment by bassamanator** [comment]
So the printer is in better physical condition, great! _I'm assuming before you did a test print in klipper, you adjusted z_offset and created a new mesh._

Given the new physical condition:
1. Klipper prints poorly
2. Marlin prints well
?

This is not a bug, because if it were a bug, Klipper would either 'go broke' or the bug would be immediately patched.

Based on what you've said, it seems that a better z_offset calibration might have taken place while Marlin was running the show. You might want to try and increase the z_offset in Klipper, you can also do this during printing:
![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/61985779/963ed39d-6219-47f7-880b-2743ca4a468d)
This happens live, and if you want to save the adjustment to the `printer.cfg`, hit save > save config, but you can save to the config after the print.

You can't directly report bugs to the Klipper repo btw, you have to do it in their discourse: 
https://klipper.discourse.group/

Post images of the marlin vs klipper prints btw.

**Comment by bassamanator** [comment]
Also wanted to mention that I have an inkling as to what the issue might be, but don't want to jump to that conclusion too quickly.

**Comment by JoernSchoenyan** [comment]
> So the printer is in better physical condition, great! _I'm assuming before you did a test print in klipper, you adjusted z_offset and created a new mesh._

Correct.

> Given the new physical condition:
> 
> 1. Klipper prints poorly
> 2. Marlin prints well
>    ?

Correct.

> This is not a bug, because if it were a bug, Klipper would either 'go broke' or the bug would be immediately patched.

Okay, maybe I should elaborate: as the Z axis using Klipper isn't moving to compensate the bed variance but does this when using Marlin stock firmware, from my understanding it has to be a bug or a configuration error from my side.

> Based on what you've said, it seems that a better z_offset calibration might have taken place while Marlin was running the show. You might want to try and increase the z_offset in Klipper

Increasing the z offset would help on the left side of the print but would be even worse on the right side. I'll try printing these tools here https://www.printables.com/de/model/42803-x-axis-level-tool to eliminate to possibility of a z axis problem (or at least make the probability smaller).

**Comment by JoernSchoenyan** [comment]
![IMG_20231129_115947](https://github.com/bassamanator/Sovol-SV06-firmware/assets/1822665/2cddb2cd-080f-4f83-bc64-b3c644754591)

This is the test print on Marlin. On Klipper it still looks like garbage with the exact same issues I have had from the beginning.

**Comment by bassamanator** [comment]
Can you type `SCREWS_TILT_CALCULATE`, and post the image that results. Just hit accept afterwards btw, you don't need to do anything.
![title-adjust](https://github.com/bassamanator/Sovol-SV06-firmware/assets/61985779/6be02acb-9ca6-480d-ba6d-7909d4894470)

Please update everything via the dashboard, and post the resulting picture *after* the update.
- Fluidd, you can do this from Settings > Software Updates.
- In Mainsail, you can do this from Machine > Update Manager.
![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/61985779/3c16ef4e-08a1-4117-b8a3-dca4bab99e0e)

Please ssh into the host and type:
1. `cd ~/printer_data/config`
2. `tree` (if you get a `command not found error`, do a `sudo apt update && sudo apt install tree`)
3. Paste the out here inside a [codeblock](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks).

Please paste both your klipper.log (you can get that from the same section where you updated) and printer.cfg at https://pastebin.com/, you do not need to create an account.

Another picture of your latest klipper mesh would also be useful.

Take your time.

PS I was thinking you have x-twist, but your very acceptable first layer in `marlin` and very unsable first layer in `klipper` complicates that theory. It's an easy thing to try but we'll get there...

**Comment by JoernSchoenyan** [comment]
The bed screws are adjusted (I've pulled out the screws on the right a tiny bit more than the tool suggests which gives me a slightly better bed mesh): 
![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/1822665/b825c02e-0a28-457d-8dc5-7573418157c9)

The current bed mesh:
![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/1822665/4d264e09-070f-4780-ad54-80ab422476cb)

The config folder:
```
pi@klipper:~/printer_data/config$ tree
.
├── cfgs
│   ├── adxl-direct.cfg
│   ├── adxl-rp2040.cfg
│   ├── adxl-rpi-pico-2x.cfg
│   ├── CALIBRATION.cfg
│   ├── kamp
│   │   ├── Adaptive_Meshing.cfg
│   │   ├── KAMP_Settings.cfg
│   │   ├── Line_Purge.cfg
│   │   ├── Smart_Park.cfg
│   │   └── Voron_Purge.cfg
│   ├── MECHANICAL_GANTRY_CALIBRATION.cfg
│   ├── misc-macros.cfg
│   ├── PARKING.cfg
│   └── TEST_SPEED.cfg
├── moonraker.conf
├── osskc.cfg
├── printer-20231118_224542.cfg
├── printer-20231118_231140.cfg
├── printer-20231118_231955.cfg
├── printer-20231118_232604.cfg
├── printer-20231118_233440.cfg
├── printer-20231119_112333.cfg
├── printer-20231119_130632.cfg
├── printer-20231119_131208.cfg
├── printer-20231119_132418.cfg
├── printer-20231119_141227.cfg
├── printer-20231119_201814.cfg
├── printer-20231119_204716.cfg
├── printer-20231122_212558.cfg
├── printer-20231122_213852.cfg
├── printer-20231124_231847.cfg
├── printer-20231125_235057.cfg
├── printer-20231126_130105.cfg
├── printer-20231128_141758.cfg
└── printer.cfg

3 directories, 34 files

```

![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/1822665/93c1cb70-09a6-4c58-a58a-5302a27d7502)

printer.cfg: https://pastebin.com/9CL9z2Qv
The klipper.log is too large for pastebin, I've uploaded it to my webspace: https://hobbygamedev.de/temp/klipper

I will now restart the machine, calibrate the z-offset and try another print to report later :-)

**Reply by JoernSchoenyan** [reply]
![IMG_20231203_150921.jpg](https://github.com/bassamanator/Sovol-SV06-firmware/assets/1822665/9246ea0f-2d8e-4bd8-b1dc-283afed57eb2)

Still very bad.

**Comment by bassamanator** [comment]
Everything looks fine. The only thing worth mentioning is that your `z_offset` is on the higher side, meaning a smaller gap between the nozzle and the bed. I don't want to discuss this but I recommend that you:
- Prepare for a probe calibration according to my `README`.
- Use two pieces of paper for the paper test: 1 thin (lined paper from school), 1 thick (printer paper).
- Try to remove history from the equation: you are calibrating this probe for the first time.

Reprint and see if that helped, also post your new `z_offset`; currently `z_offset = 1.985`.

If this doesn't help, then I recommend following the pinned discussion on `x twist`. Your config is slightly out of date, so update that first to get the axis twist compensation section. 
![image](https://github.com/bassamanator/Sovol-SV06-firmware/assets/61985779/554d6f0e-102a-4517-a866-1019d9efa966)

**Comment by bassamanator** [comment]
Good to go now?

**Comment by JoernSchoenyan** [comment]
To be honest: no, I am so lost in the Klipper world that I want to use Marlin again for a while and come back to Klipper later to not loose my fun in 3D printing.

I can confirm that ABL is doing something, I just couldn't see it. But I made a small video and could see it there. I still can't get why the print is that bad, especially on the left side of the bed as the result looks not like I would expect if I compare it with the bed mesh.

**Comment by bassamanator** [comment]
As I say in the preface to the repo, "Klipper is not a must". And I understand your frustration more than most.

### Where you are, I once was.
![calibration_prints](https://github.com/bassamanator/Sovol-SV06-firmware/assets/61985779/4354b844-8659-45be-976f-60603f6ee94b)
This is only a partial depiction. I checked my Klipper print history and I have over hundred of these prints (I'm not exaggerating).

Feel free to re-open the discussion.

Happy printing.

**Comment by JoernSchoenyan** [comment]
After taking a break, I wanted to give Klipper another chance and I can report about my success!

It was indeed x-twist. I'm not 100% sure, but I think when the printer was delivered, this issue was not existent on my printer. My guess is, that the injection molded plastic that holds the z-axis motors degrades over time. The motor on the left was just slightly tilted, but the motor on the right was tilted a lot.

My solution was - for now - a small shim from Printables together with Klipper's axis twist compensation (I needed a feeler gauge for that, which I now consider as essential). I did a test print which was very good now! Thank you very much!
